### PR TITLE
Testing : Add 'coding : UTF-8' test in code quality test

### LINKED ISSUE #16257: Ensuring that sympy is importable on Python 2
#16254 shows that it's possible to pass the Travis tests even if sympy cannot be imported under Python 2. See [https://github.com/sympy/sympy/issues/16254#issuecomment-472813993](https://github.com/sympy/sympy/issues/16254#issuecomment-472813993)

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #16257

#### Brief description of what is fixed or changed

Current test suite does not properly detect import error for python 2, as the reason described in #16257

I have found the regular expression in
https://www.python.org/dev/peps/pep-0263/#defining-the-encoding
to match the encoding,

And used the trick from
https://www.quora.com/How-do-you-check-to-see-if-a-string-in-Python-is-in-ASCII
to detect unicode character in the string.

#### Other comments

This may be deleted after deprecating python 2.7

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/utilities/tests/test_code_quality.py [hunk 1]
```diff
@@ -184,6 +199,19 @@ def test_this_file(fname, test_file):
             if func_is_re.search(line) and not test_file_re.search(fname):
                 assert False, message_func_is % (fname, idx + 1)
 
+            # XXX Python 2 unicode import test.
+            # May remove after deprecating python 2.7.
+            match = encoding_header_re.match(line)
```
Comment (jksuom) [inline sympy/utilities/tests/test_code_quality.py]: Would it suffice to check this only for the first two lines?

File: sympy/utilities/tests/test_code_quality.py [hunk 2]
```diff
@@ -184,6 +199,20 @@ def test_this_file(fname, test_file):
...
                 assert False, message_func_is % (fname, idx + 1)
 
+            # XXX Python 2 unicode import test.
+            # May remove after deprecating python 2.7.
+            if idx in (0, 1):
+                match = encoding_header_re.match(line)
+                if match:
+                    if match[1].lower() == 'utf-8':
+                        has_coding_utf8 = True
+            try:
```
Comment (jksuom) [inline sympy/utilities/tests/test_code_quality.py]: How about something like this:

    if not has_coding_utf8:
        try:
            line.encode(encoding='ascii')
        except UnicodeEncodeError:
            assert False, message_py2_unicode % (fname, idx + 1)

**Comment by sylee957** [comment]
I'm not sure there exists a good way to test the testing module itself.
Though, I believe there is, but I see most of the SymPy testing module doesn't come up with testing itself.

I have added a dummy file though.

**Comment by jksuom** [comment]
The `line.encode` tests will probably take some time. Would it be reasonable to run them only on Python 2?

**Comment by oscarbenjamin** [comment]
I think that even after dropping support for Python 2 we should disallow non-ASCII in the source code as a matter of code quality. The only cases I know of where it can be necessary are in the `test_` files that verify unicode input/output (e.g. in printing). Doctests use `unicode=False` for precisely this reason.

I think it would be better to have a whitelist of which files are allowed to use non-ASCII. Then this test should verify that all the non-whitelisted files are ASCII (regardless of encoding cookie) and that the whitelisted files have the encoding cookie.

**Comment by sylee957** [comment]
@oscarbenjamin 
I see your idea about whitelisting, but in my opinion, whitelisting may need some additional management, and in current codebase, the slowdown isn't much great that it would need much manual work.

I only see it increased from 5.33 seconds to 31.263 seconds in my [build](https://travis-ci.org/sympy/sympy/builds/507434341)

And even if we reach 1 million lines, or even more, it can be marked `@slow` and can be managed under separate stage, so it won't exceed any timeout.

However, inversely, I find that some files are containing `# -*- coding: utf-8 -*-` in unnecessary way, like
https://github.com/sympy/sympy/blob/f3c556d10ccbbea997ba4dc3b8a66f32d1a1ea5f/sympy/codegen/tests/test_approximations.py
but it does not have any unicode in real.

Considering that some editors are automatically appending those lines as dummy, I think the code quality test may have to fail for files containing unnecessary lines like those.

**Comment by oscarbenjamin** [comment]
I'm not proposing to use a whitelist for performance reasons.

Adding an encoding cookie makes a utf-8 encoded file containing non-ASCII importable under Python 2. Under Python 3 the file will be importable anyway. Of course we want all our files to be importable under Python 2 so we should ensure that the cookies are used when needed.

However this is a separate question from whether we want to have ASCII characters in the source. I think we do not. So the quality test should reject any file containing non-ASCII characters. There are a small number of files that have a good reason for doing so though. So those files should have an encoding cookie and be whitelisted in the quality tests.

**Comment by sylee957** [comment]
Okay, I've added brief whitelisting feature, though I have just briefly added some patterns to make the test pass.

**Comment by sylee957** [comment]
Okay, I had pinpointed each files,
so anyone who want to contribute, can get the work easily verified.

Though, I have some other ideas, that we should also forbid
`# coding=utf-8` in files, if there is no unicode.

**Comment by oscarbenjamin** [comment]
Is there a list of files to fix somewhere?

**Comment by sylee957** [comment]
Check ‘’unicode_whitelist’’

**Comment by oscarbenjamin** [comment]
Those would probably make good easy to fix issues

**Comment by asmeurer** [comment]
I'm not sure I agree that we should ban non-ASCII characters. Something as simple as having someone's name that is spelled with an accent requires a non-ASCII character.

Can you provide a list of which files contain non-ASCII characters, and why? The only one I know about is `test_pretty.py` (and `AUTHORS`, but that isn't part of the source).

**Comment by oscarbenjamin** [comment]
I believe the code quality guideline for the Python stdlib is that code should be kept ASCII apart from author names (in comments).

File: sympy/utilities/tests/test_code_quality.py [hunk 3]
```diff
@@ -37,6 +38,51 @@
...
+    r'*/combinatorics/tests/test_coset_table.py',
+    r'*/crypto/crypto.py',
+    r'*/discrete/transforms.py',
+    r'*/liealgebras/root_system.py',
+    r'*/liealgebras/type_g.py',
+    r'*/liealgebras/weyl_group.py',
+    r'*/liealgebras/tests/test_type_G.py',
+    r'*/ntheory/residue_ntheory.py',
+    r'*/physics/mechanics/rigidbody.py',
+    r'*/polys/subresultants_qq_zz.py'
```
Comment (asmeurer) [inline sympy/utilities/tests/test_code_quality.py]: Oh I see these are the files. This file in particular seems to be using curly quotes. Those could be replaced with regular quotes (it's up to the doc renderer to render curly quotes for typography). 

discrete/transforms and residue_ntheory use "Möbius" in some of the docstrings. That seems fine to me. 

The liealgebra files use greek symbols in docstrings, which could be replaced with latex. 

rigitbody.py uses `H = I·w + r×Mv`, which could be redone in latex.

crypto.py has some en-dashes, which could be replaced with regular dashes, and "Vigenère", which is fine.

free_groups.py has an en-dash which should clearly be a regular dash (it's in a LaTeX equation)

test_coset_table.py has some subscript numbers in comments (like `₁`), which could be replaced with normal numbers

**Comment by asmeurer** [comment]
We could whitelist allowed characters outside of the printing files. It could be useful to avoid invisible characters like BOM (https://github.com/sympy/sympy/issues/10974) or invisible ASCII characters (I think we already check for tabs and carriage returns).

**Comment by sylee957** [comment]
I see the slowdown is not very critical on travis, but in some of my old computers, it takes more than 100 seconds.
However, there would not much an option to check unicode character instead of scanning all the strings in the codebase.

@asmeurer Do you mean that there should also be some additional whitelist for allowed characters?

**Comment by asmeurer** [comment]
@oscarbenjamin if you merge this with the 1.5 branch, there is an error

```
Traceback (most recent call last):
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 299, in test_this_file_encoding
    line.encode(encoding='ascii')
UnicodeEncodeError: 'ascii' codec can't encode character '\xa0' in position 5: ordinal not in range(128)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 346, in test_files
    check_directory_tree(SYMPY_PATH, test, exclude)
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 155, in check_directory_tree
    check_files(glob(join(root, pattern)), file_check, exclusions)
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 171, in check_files
    file_check(fname)
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 194, in test
    test_this_file_encoding(fname, test_file)
  File "/home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_code_quality.py", line 302, in test_this_file_encoding
    message_py2_unicode_not_whitelisted % (fname, idx + 1)
AssertionError: File contains a unicode character : /home/aaronmeurer/Documents/python/sympy/sympy/sympy/utilities/tests/test_pickling.py, line 398. But with no encoding header. See https://www.python.org/dev/peps/pep-0263/ and add '# coding=utf-8'
```

There is a Unicode character in that file that should be removed. 

There is a similar error that recently snuck into master. I'm not really clear why the tests didn't fail on Travis. master currently cannot be imported in Python 2 on my computer. 

We should probably get this PR merged into the release so that we don't accidentally release an unimportable version of SymPy. It might also be worth investigating why sympy imports at all in Travis Python 2 with the current master.

**Comment by asmeurer** [comment]
Also I would add that in both the cases of errors in 1.5 and master, the Unicode character should actually be an ASCII character. In test_pickling.py there is a non-breaking space, and in solvers.py in master there is an em dash instead of `-`. So the error message should tell people to check for Unicode characters on the given line, and see if they are supposed to be ASCII. Only if the Unicode characters is intentional should a coding comment be added. 

And even for many intentional cases we should avoid them. For example, greek characters should be spelled out, math in docstrings should use LaTeX, and many names with accents should be transliterated (like Groebner, although I think having accents in names is fine if that is the correct spelling of them). For one thing, a single Unicode character in a file will mask any invalid ones, like the one for the error above. Perhaps the whitelist should be more fine grained.

**Comment by asmeurer** [comment]
There is an issue with the logic here. You get the error

```
AssertionError: File contains a unicode encoding header : /home/aaronmeurer/Documents/python/sympy/sympy/sympy/physics/wigner.py, line 1. But it would not be needed.
```

but if you remove the header, it fails with

```
AssertionError: File contains a unicode character : /home/aaronmeurer/Documents/python/sympy/sympy/sympy/physics/wigner.py, line 785. But with no encoding header. See https://www.python.org/dev/peps/pep-0263/ and add '# coding=utf-8'
```

(The file has doctests that use `use_unicode=True`, which I'm not sure if we should be using, but that's another discussion)

**Comment by asmeurer** [comment]
> (The file has doctests that use use_unicode=True, which I'm not sure if we should be using, but that's another discussion)

I think we shouldn't at least until https://github.com/sympy/sympy/issues/15700 is fixed.

**Comment by oscarbenjamin** [comment]
I'm not sure what has changed in this PR that causes `bin/test_executable` to run in the optional dependency tests. It has picked up two files in sympy/physics/optics.

**Comment by oscarbenjamin** [comment]
I intend to merge this as #18005 into the 1.5 branch. Please comment there if you have suggestions for changes. This will not need to be merged into master because 1.5 will soon be merged back into master.

**Comment by sylee957** [comment]
Okay, I'll close this since it's merged in #18005.

**Pull request closed without merging**